librosa/core/pitch.py

```diff
@@ -333,3 +338,446 @@ def piptrack(y=None, sr=22050, S=None, n_fft=2048, hop_length=None,
+        the edges of the signal ``y``. By default (``pad_mode="reflect"``),
+        ``y`` is padded on both sides with its own reflection, mirrored around
+        its first and last sample respectively.
+        If ``center=False``,  this argument is ignored.
+        .. see also:: `np.pad`
+
+    Returns
+    -------
+    f0: np.ndarray [shape=(n_frames,)]
+        time series of fundamental frequencies in Hertz.
+
+    voiced_flag: np.ndarray [shape=(n_frames,)]
+        time series containing boolean flags indicating whether a frame is voiced or not.
+
+    voiced_prob: np.ndarray [shape=(n_frames,)]
+        time series containing the probability that a frame is voiced.
+
+    See Also
+    --------
+    librosa.yin
+        Fundamental frequency (F0) estimation using the YIN algorithm.
+
+    Examples
+    --------
+    Computing a fundamental frequency (F0) curve from an audio input
+
+    >>> y, sr = librosa.load(librosa.ex('trumpet'))
+    >>> f0, voiced_flag, voiced_probs = librosa.pyin(y, fmin=librosa.note_to_hz('C2'), fmax=librosa.note_to_hz('C7'))
+    >>> times = librosa.times_like(f0)
+
+
+    Overlay F0 over a spectrogram
+
+    >>> import matplotlib.pyplot as plt
+    >>> D = librosa.amplitude_to_db(np.abs(librosa.stft(y)), ref=np.max)
+    >>> fig, ax = plt.subplots()
+    >>> img = librosa.display.specshow(D, x_axis='time', y_axis='log', ax=ax)
+    >>> ax.set(title='pYIN fundamental frequency estimation')
+    >>> fig.colorbar(img, ax=ax, format="%+2.f dB")
+    >>> ax.plot(times, np.where(voiced_flag, f0, np.nan), label='f0', color='w')  # mask unvoiced frames with np.nan
+    >>> ax.legend(loc='upper right')
+    '''
+    if win_length > frame_length:
+        raise ParameterError('win_length cannot exceed frame_length')
+
+    # Set the default hop if it is not already specified.
+    if hop_length is None:
+        hop_length = int(frame_length // 4)
+
+    # Check that audio is valid.
+    util.valid_audio(y, mono=True)
+
+    # Pad the time series so that frames are centered
+    if center:
+        y = np.pad(y, frame_length // 2, mode=pad_mode)
+
+    # Frame audio.
+    y_frames = util.frame(y, frame_length=frame_length, hop_length=hop_length)
+
+    # Calculate minimum and maximum periods
+    min_period = max(math.floor(sr / fmax), 2)
+    max_period = min(math.ceil(sr / fmin), frame_length-win_length-1)
+
+    # Calculate cumulative mean normalized difference function.
+    yin_frames = _cumulative_mean_normalized_difference(
+        y_frames, frame_length, win_length, min_period, max_period)
+
+    # Parabolic interpolation.
+    parabolic_shifts = _parabolic_interpolation(yin_frames)
+
+    # Find Yin candidates and probabilities.
+    # The implementation here follows the official pYIN software which
+    # differs from the method described in the paper.
+    # 1. Define the prior over the thresholds.
+    thresholds = np.linspace(0, 1, n_thresholds + 1)
```

Comment:
why n_thresholds +1?  should we include the 1 endpoint here?

Reply:
On line 708 we need `p(a <= x < b)` over each threshold interval. I've implemented this by taking the cdf of the beta distribution at the endpoints of each interval and taking differences (so we need both endpoints). Later, on line 723, the first endpoint is discarded but we still need to check if the troughs are below `1`.